File: modules/react-native-nitro-recycler-list/android/src/main/java/com/margelo/nitro/recyclerlist/RecyclerListRefreshEventState.kt
package com.margelo.nitro.recyclerlist

internal data class RecyclerListRefreshSnapshot(
  val phase: NativeRefreshPhase,
  val offset: Double,
  val progress: Double,
)

/** 统一规范化刷新事件，并将高频位移与低频阶段通知分流。 */
internal class RecyclerListRefreshEventState {
  var phase: NativeRefreshPhase = NativeRefreshPhase.IDLE
    private set

  fun publish(
    nextPhase: NativeRefreshPhase,
    offset: Double,
    progress: Double,
    onPull: (RecyclerListRefreshSnapshot) -> Unit,
    onPhaseChanged: (NativeRefreshPhase) -> Unit,
  ) {
    onPull(
      RecyclerListRefreshSnapshot(
        nextPhase,
        offset.coerceAtLeast(0.0),
        progress.coerceIn(0.0, 1.0),
      ),
    )
    if (phase != nextPhase) {
      phase = nextPhase
      onPhaseChanged(nextPhase)
    }
  }
}

File: modules/react-native-nitro-recycler-list/ios/RecyclerListRefreshEventState.swift
import Foundation

struct RecyclerListRefreshSnapshot {
  let phase: NativeRefreshPhase
  let offset: Double
  let progress: Double
}

/// 统一规范化刷新事件，并将高频位移与低频阶段通知分流。
final class RecyclerListRefreshEventState {
  private(set) var phase: NativeRefreshPhase = .idle
  private(set) var offset: Double = 0

  func publish(
    phase nextPhase: NativeRefreshPhase,
    offset nextOffset: Double,
    progress nextProgress: Double,
    onPull: (RecyclerListRefreshSnapshot) -> Void,
    onPhaseChanged: (NativeRefreshPhase) -> Void
  ) {
    offset = max(0, nextOffset)
    onPull(
      RecyclerListRefreshSnapshot(
        phase: nextPhase,
        offset: offset,
        progress: min(1, max(0, nextProgress))
      )
    )
    if phase != nextPhase {
      phase = nextPhase
      onPhaseChanged(nextPhase)
    }
  }
}
